crates/task-impls/src/quorum_proposal.rs

```diff
@@ -80,8 +71,160 @@ struct ProposalDependencyHandle<TYPES: NodeType> {
+        commit_and_metadata: CommitmentAndMetadata<TYPES::BlockPayload>,
+    ) -> bool {
+        if self.quorum_membership.get_leader(view) != self.public_key {
+            // This is expected for view 1, so skipping the logging.
+            if view != TYPES::Time::new(1) {
+                error!(
+                    "Somehow we formed a QC but are not the leader for the next view {:?}",
+                    view
+                );
+            }
+            return false;
+        }
+
+        let consensus = self.consensus.read().await;
+        let parent_view_number = &consensus.high_qc.get_view_number();
+        let mut reached_decided = false;
+
+        let Some(parent_view) = consensus.validated_state_map.get(parent_view_number) else {
+            // This should have been added by the replica?
+            error!("Couldn't find parent view in state map, waiting for replica to see proposal\n parent view number: {}", **parent_view_number);
+            return false;
+        };
+        // Leaf hash in view inner does not match high qc hash - Why?
+        let Some((leaf_commitment, state)) = parent_view.get_leaf_and_state() else {
+            error!(
+                ?parent_view_number,
+                ?parent_view,
+                "Parent of high QC points to a view without a proposal"
+            );
+            return false;
+        };
+        if leaf_commitment != consensus.high_qc.get_data().leaf_commit {
+            // NOTE: This happens on the genesis block
+            debug!(
+                "They don't equal: {:?}   {:?}",
+                leaf_commitment,
+                consensus.high_qc.get_data().leaf_commit
+            );
+        }
+        let Some(leaf) = consensus.saved_leaves.get(&leaf_commitment) else {
+            error!("Failed to find high QC of parent.");
+            return false;
+        };
+        if leaf.get_view_number() == consensus.last_decided_view {
+            reached_decided = true;
+        }
+
+        let parent_leaf = leaf.clone();
+
+        let original_parent_hash = parent_leaf.commit();
+
+        let mut next_parent_hash = original_parent_hash;
+
+        // Walk back until we find a decide
+        if !reached_decided {
+            debug!(
+                "We have not reached decide from view {:?}",
+                self.view_number
+            );
+            while let Some(next_parent_leaf) = consensus.saved_leaves.get(&next_parent_hash) {
+                if next_parent_leaf.get_view_number() <= consensus.last_decided_view {
+                    break;
+                }
+                next_parent_hash = next_parent_leaf.get_parent_commitment();
+            }
+            debug!("updated saved leaves");
+            // TODO do some sort of sanity check on the view number that it matches decided
+        }
+
+        // Special case: if we have a decided upgrade certificate AND it does not apply a version to the current view, we MUST propose with a null block.
```

Comment:
Is this a TODO?

Reply:
Copied from consensus task, so probably yes but I think it might be an artifact from before.

Reply:
We can probably get rid of this TODO since the dependency task makes sure the view number always matches.